crates/matrix-sdk-ui/src/timeline/controller/read_receipts.rs

```diff
@@ -273,27 +286,56 @@ struct FullReceipt<'a> {
 /// A read receipt update in the timeline.
 #[derive(Clone, Debug, Default)]
 struct ReadReceiptTimelineUpdate {
+    /// The position of the old timeline item that had the receipt of the user,
+    /// if any.
+    old_item_pos: Option<usize>,
     /// The old event that had the receipt of the user, if any.
     old_event_id: Option<OwnedEventId>,
+    /// The position of the new timeline item that has the receipt of the user,
+    /// if any.
+    new_item_pos: Option<usize>,
     /// The new event that has the receipt of the user, if any.
     new_event_id: Option<OwnedEventId>,
 }
 
 impl ReadReceiptTimelineUpdate {
     /// Remove the old receipt from the corresponding timeline item.
-    fn remove_old_receipt(&self, items: &mut ObservableItemsTransaction<'_>, user_id: &UserId) {
+    ///
+    /// Return the position of the updated timeline item.
+    fn remove_old_receipt(&mut self, items: &mut ObservableItemsTransaction<'_>, user_id: &UserId) {
         let Some(event_id) = &self.old_event_id else {
             // Nothing to do.
             return;
         };
 
-        let Some((receipt_pos, event_item)) = rfind_event_by_id(items, event_id) else {
-            debug!(%event_id, %user_id, "inconsistent state: old event item for read receipt was not found");
-            return;
+        let item_pos = match &self.old_item_pos {
+            Some(item_pos) => *item_pos,
+            None => {
+                let Some(old_pos) = items
```

Comment:
here and below, light style suggestion: I think this would read slightly simpler as `self.old_item_pos.or_else(|| /* find with linear search */)` and then an `let Some(item_pos) = … else { debug!(); return;`, as it avoids nesting several levels of depth.

Reply:
Indeed, it's more compact.